code/datums/elements/tool_renaming.dm

```diff
@@ -0,0 +1,68 @@
+/**
+ * Renaming tool element
+ *
+ * When using this tool on an object with UNIQUE_RENAME,
+ * lets the user rename/redesc it.
+ */
+/datum/element/tool_renaming
+
+/datum/element/tool_renaming/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ITEM_INTERACTING_WITH_ATOM, PROC_REF(attempt_rename))
+
+/datum/element/tool_renaming/Detach(datum/source)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ITEM_INTERACTING_WITH_ATOM)
+
+/datum/element/tool_renaming/proc/attempt_rename(datum/source, mob/living/user, atom/interacting_with, list/modifiers)
+	if(!isobj(interacting_with))
+		return NONE
+	
+	var/obj/renamed_obj = interacting_with
+
+	if(!(renamed_obj.obj_flags & UNIQUE_RENAME))
+		return NONE
+
+	INVOKE_ASYNC(src, PROC_REF(async_rename), user, renamed_obj)
+	return ITEM_INTERACT_SUCCESS
+
+/datum/element/tool_renaming/proc/async_rename(mob/living/user, obj/renamed_obj)
+	var/pen_choice = tgui_input_list(user, "What would you like to edit?", "Pen Setting", list(OPTION_RENAME, OPTION_DESCRIPTION, OPTION_RESET))
+	if(QDELETED(renamed_obj) || !user.can_perform_action(renamed_obj) || isnull(pen_choice))
+		return
+
+	switch(pen_choice)
+		if(OPTION_RENAME)
+			var/input = tgui_input_text(user, "What do you want to name [renamed_obj]?", "Object Name", "[renamed_obj.name]", MAX_NAME_LEN)
+			var/old_name = renamed_obj.name
```

Comment:
let's save some var-accesses with a little bit of code usage (and just make everything nice and explicit as possible)

```suggestion
			var/old_name = renamed_obj.name
			var/input = tgui_input_text(user, "What do you want to name [renamed_obj]?", "Object Name", old_name, MAX_NAME_LEN)
```

(we don't need to double stringify `old_name` here because `name` is already a string)